[Workers] Document streams_byob_reader_detaches_buffer compatibility flag (#2644)
cloudflare/cloudflare-docs: products/workers/src/content/platform/compatibility-dates.md

Newest changes are listed first.

### Streams BYOB reader detaches buffer

<table><tbody>
  <tr><td><strong>Default as of</strong></td><td>2021-11-10</td></tr>
  <tr><td><strong>Flag to enable early</strong></td><td><code>streams_byob_reader_detaches_buffer</code></td></tr>
  <tr><td><strong>Flag to disable</strong></td><td><code>streams_byob_reader_does_not_detach_buffer</code></td></tr>
</tbody></table>

> Originally, the Workers runtime did not detach the ArrayBuffers from user-provided TypedArrays when using the [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods), as required by the Streams spec, meaning it was possible to inadvertently reuse the same buffer for multiple `read()` calls. This change brings us in line with the spec.

Reviewer: ```suggestion
Originally, the Workers runtime did not detach the `ArrayBuffer`s from user-provided TypedArrays when using the [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods), as required by the Streams spec, meaning it was possible to inadvertently reuse the same buffer for multiple `read()` calls. This change brings us in line with the spec.
```

> Originally, the Workers runtime did not detach the ArrayBuffers from user-provided TypedArrays when using the [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods), as required by the Streams spec, meaning it was possible to inadvertently re-use the same buffer for multiple `read()` calls. This change brings us in line with the spec.

Reviewer: ```suggestion
Originally, the Workers runtime did not detach the ArrayBuffers from user-provided TypedArrays when using the [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods), as required by the Streams spec, meaning it was possible to inadvertently reuse the same buffer for multiple `read()` calls. This change brings us in line with the spec.
```

> User code should never try to re-use an ArrayBuffer that has been passed in to a [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods). The more recently added extension method `readAtLeast()` will always detach the ArrayBuffer and is unaffected by this feature flag setting.

Reviewer: ```suggestion
User code should never try to reuse an `ArrayBuffer` that has been passed into a [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods). The more recently added extension method `readAtLeast()` will always detach the `ArrayBuffer` and is unaffected by this feature flag setting.
```

> User code should never try to reuse an `ArrayBuffer` that has been passed into a [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods). The more recently added extension method `readAtLeast()` will always detach the `ArrayBuffer` and is unaffected by this feature flag setting.

Reviewer 2: I think you can reuse the buffer, just not under the same reference? That is, when `read()` returns, it gives you a new reference to the buffer, which you can then use in the next call. This seems like the whole point of BYOB, to avoid allocation overhead, so we should probably be a bit clearer here.

Author: I added a little example.

Newest changes are listed first.

### Streams BYOB reader detaches buffer

<table><tbody>
  <tr><td><strong>Default as of</strong></td><td>2021-11-10</td></tr>
  <tr><td><strong>Flag to enable early</strong></td><td><code>streams_byob_reader_detaches_buffer</code></td></tr>
  <tr><td><strong>Flag to disable</strong></td><td><code>streams_byob_reader_does_not_detach_buffer</code></td></tr>
</tbody></table>

> Originally, the Workers runtime did not detach the `ArrayBuffer`s from user-provided TypedArrays when using the [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods), as required by the Streams spec, meaning it was possible to inadvertently reuse the same buffer for multiple `read()` calls. This change brings us in line with the spec.

Reviewer: ```suggestion
Originally, the Workers runtime did not detach the `ArrayBuffer`s from user-provided TypedArrays when using the [BYOB reader's `read()` method](/runtime-apis/streams/readablestreambyobreader#methods), as required by the Streams spec, meaning it was possible to inadvertently reuse the same buffer for multiple `read()` calls. This change makes Workers conform to the spec.
```
